wkcuber/downsampling_utils.py

```diff
@@ -0,0 +1,351 @@
+import logging
+import math
+from enum import Enum
+from itertools import product
+from typing import Optional, Tuple, List, Callable, cast
+
+import numpy as np
+from scipy.ndimage import zoom
+from wkw import wkw
+
+from wkcuber.api.View import View
+from wkcuber.mag import Mag
+from wkcuber.utils import time_start, time_stop
+
+
+class InterpolationModes(Enum):
+    MEDIAN = 0
+    MODE = 1
+    NEAREST = 2
+    BILINEAR = 3
+    BICUBIC = 4
+    MAX = 5
+    MIN = 6
+
+
+DEFAULT_EDGE_LEN = 256
+
+
+def use_logging(
+    offset: Tuple[int, int, int],
+    size: Tuple[int, int, int],
+    job_count_per_log: int,
+    global_offset: Tuple[int, int, int],
+    num_chunks_per_dim: Tuple[int, int, int],
+) -> bool:
+    # calculate if a specific chunk should log its results
+    chunk_xzy_idx = (np.array(offset) - np.array(global_offset)) // size
+    i = (
+        chunk_xzy_idx[2] * num_chunks_per_dim[0] * num_chunks_per_dim[1]
+        + chunk_xzy_idx[1] * num_chunks_per_dim[1]
+        + chunk_xzy_idx[0]
+    )
+
+    return i % job_count_per_log == 0
+
+
+def determine_buffer_edge_len(dataset: wkw.Dataset) -> int:
+    if hasattr(dataset.header, "file_len") and hasattr(dataset.header, "block_len"):
```

Comment:
This probably cannot happen, but conceptionally only one of these attributes could be set (e.g., block_len == 1). file_len would still default to 32 in which case the effective cube_length will be 32. However, the current code wouldn't take that length into account (since the header would need to have both attributes defined) and simply return the default edge length.

From my understanding, the header always has both properties defined (in the wkw library, they simply default to 32). So, the actual question is probably: why do we need the `hasattr` calls here? If it's only to satisfy mypy, we could replace this with `assert hasattr(dataset.header,  "file_len") and ...`.

Reply:
I haven't push the latest commit. The type of dataset is actually `Union[wkw.Dataset, TiffMag]`. `TiffMag` does not have those attributes

Reply:
Ah, I see. Can you do an is-instance-of check and differentiate via an if-else between the two cases? Might be a tiny bit cleaner.